packages/playwright-core/src/androidServerImpl.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { LaunchAndroidServerOptions } from './client/types';
+import { ws } from './utilsBundle';
+import type { WebSocketEventEmitter } from './utilsBundle';
+import type { BrowserServer } from './client/browserType';
+import { createGuid } from './utils';
+import { createPlaywright } from './server/playwright';
+import { PlaywrightServer } from './remote/playwrightServer';
+
+export class AndroidServerLauncherImpl {
+  async launchServer(options: LaunchAndroidServerOptions = {}): Promise<BrowserServer> {
+    const playwright = createPlaywright('javascript');
+    // 1. Pre-connect to the device
+    let devices = await playwright.android.devices({
```

Comment:
I'd avoid a long-living connection here and would create a device upon the connection. I.e. instead of `preLaunchedAndroidDevice: device` pass `preLaunchedAndroidDevice: deviceFactory`.

Reply:
Good idea but this makes it harder when a device gets disconnected during launchServer and connect(). With the current impl the ws server will shut down when the adb connection drops.